### ISSUE TITLE
Book recommendation about Kestrel architecture

### ISSUE THREAD

**Issue by Rick-Anderson** [dotnet/AspNetCore.Docs#28378]
### Discussed in https://github.com/dotnet/AspNetCore.Docs/discussions/28068

<div type='discussions-op-text'>

<sup>Originally posted by **lukasgmt** January 11, 2023</sup>
Could you recommend any deep dive book that describes the internal Kestrel architecture. Most of the books focus on the ASP.NET, MVC, Razor etc. while I am mostly focused on all the available middlewares and their configurations. 

What I am attempting to write is a reverse proxy that:
* Is a mitigator of OAuth authentication
* Hides OAuth tokens behind cookies
* Refresh them inside my reverse proxy depending on how recently the tokens were resolved (in a timer)
* Store tokens in Redis in order to share them across replicas of the reversed proxy

I am trying to use embedded Cookie and Authentication middlewares but so far I am failing in properly configuring them.

Would you recommend to implement it on my own or try to reuse some of OAuth + Cookies middeware?</div>

---
[Associated WorkItem - 63807](https://dev.azure.com/msft-skilling/Content/_workitems/edit/63807)

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: f22dcddb-0c25-be1a-9a02-9e71dce86bdb
* Version Independent ID: 3e0e12c2-e544-5655-e001-d2ec180b11a8
* Content: [Kestrel web server implementation in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-7.0)
* Content Source: [aspnetcore/fundamentals/servers/kestrel.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/fundamentals/servers/kestrel.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

**Comment by Rick-Anderson** [comment]
@Tratcher  what's the best way to deep dive into Kestrel. Let me know if your answer should be added to this doc.

**Comment by Tratcher** [comment]
Our docs are the most extensive resource I'm aware of for Kestrel. However, I don't think the internals are that important for your scenario. Have you looked into using YARP? It's a reverse proxy we built on top of AspNetCore/Kestrel. https://github.com/microsoft/reverse-proxy

Here are some examples I found that build something similar using YARP: 
https://github.com/manfredsteyer/yarp-auth-proxy
https://damienbod.com/2021/01/11/protecting-legacy-apis-with-an-asp-net-core-yarp-reverse-proxy-and-azure-ad-oauth/
